### PR TITLE
Standardization of GT5u Chemical Recipes

### PR DESCRIPTION
**Pull request by GlodBlock**
continue work of https://github.com/GTNewHorizons/GT5-Unofficial/pull/983

all chemical thing in gt5 is done now.
please don't merge until these places are all finished:
coremod: https://github.com/GTNewHorizons/NewHorizonsCoreMod/pull/327
bartworks: https://github.com/GTNewHorizons/bartworks/pull/144
goodgenerator: https://github.com/GTNewHorizons/GoodGenerator/pull/49
gtpp: https://github.com/GTNewHorizons/GTplusplus/pull/223
gtnh-lanth: https://github.com/GTNewHorizons/GTNH-Lanthanides/pull/16
computronics: https://github.com/GTNewHorizons/Computronics/pull/17

### CODE REVIEW

File: src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java [hunk 1]
```diff
@@ -3714,36 +3809,35 @@ private void addPyrometallurgicalRecipes() {
...
-        GT_Values.RA.addBlastRecipe(Materials.Chalcopyrite.getDust(1), Materials.Quartzite.getDust(2), Materials.Oxygen.getGas(3000), Materials.SulfurDioxide.getGas(2000), Materials.CupricOxide.getDust(1), Materials.Ferrosilite.getDust(1), 120, 120, 1200);
-        GT_Values.RA.addBlastRecipe(Materials.Chalcopyrite.getDust(1), Materials.CertusQuartz.getDust(1), Materials.Oxygen.getGas(3000), Materials.SulfurDioxide.getGas(2000), Materials.CupricOxide.getDust(1), Materials.Ferrosilite.getDust(1), 120, 120, 1200);
 
-        GT_Values.RA.addBlastRecipe(Materials.Pyrite.getDust(1), GT_Values.NI, Materials.Oxygen.getGas(3000), Materials.SulfurDioxide.getGas(2000), Materials.BandedIron.getDust(1), Materials.Ash.getDustTiny(1), 120, 120, 1200);
+        GT_Values.RA.addBlastRecipe(Materials.Tetrahedrite.getDust(1), GT_Values.NI, Materials.Oxygen.getGas(6000), Materials.SulfurDioxide.getGas(2000), Materials.CupricOxide.getDust(6), Materials.AntimonyTrioxide.getDustTiny(15), 120, 120, 1200);
```
[the remark this answers is not in the document]
Comment (GlodBlock) [reply src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java:3884]: it is tiny dust

File: src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java [hunk 2]
```diff
@@ -3792,39 +3884,47 @@ private void addPyrometallurgicalRecipes() {
...
+            	//Potassium Nitride
+                //K + HNO3 = KNO3 + H
+            	GT_Values.RA.addChemicalRecipe(Materials.Potassium.getDust(1), GT_Utility.getIntegratedCircuit(1), Materials.NitricAcid.getFluid(1000), Materials.Hydrogen.getGas(1000), Materials.PotassiumNitrade.getDust(5), 100, 30);
+
+            	//Chromium Trioxide
+                //CrO2 + O = CrO3
+            	GT_Values.RA.addChemicalRecipe(Materials.ChromiumDioxide.getDust(3), GT_Utility.getIntegratedCircuit(1), Materials.Oxygen.getGas(1000), GT_Values.NF, Materials.ChromiumTrioxide.getDust(4), GT_Values.NI,100, 60);
+
+            	//Potassium Dichromate
+                //2KNO3 + 2CrO3 = K2Cr2O7 + NO
```
[the remark this answers is not in the document]
Comment (GlodBlock) [reply src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java]: the basic chemical reactor is too small to hold all products, so i remove the oxygen output in basic chemical reactor, and put bacl in lcr.

File: src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java [hunk 3]
```diff
@@ -528,24 +528,26 @@ public void run() {
...
-        GT_Values.RA.addElectrolyzerRecipe(GT_Values.NI, GT_Values.NI, new FluidStack(ItemList.sLeadZincSolution, 8000), Materials.Water.getFluid(2000), GT_OreDictUnificator.get(OrePrefixes.dust, Materials.Lead, 1), GT_OreDictUnificator.get(OrePrefixes.dust, Materials.Silver, 1), GT_OreDictUnificator.get(OrePrefixes.dust, Materials.Zinc, 1), GT_OreDictUnificator.get(OrePrefixes.dust, Materials.Sulfur, 3), null, null, null, 300, 192);
+
+        GT_Values.RA.addElectrolyzerRecipe(GT_Values.NI, GT_Values.NI, new FluidStack(ItemList.sLeadZincSolution, 8000), Materials.Water.getFluid(2000), GT_OreDictUnificator.get(OrePrefixes.dust, Materials.Lead, 1), GT_OreDictUnificator.get(OrePrefixes.dust, Materials.Silver, 1), GT_OreDictUnificator.get(OrePrefixes.dust, Materials.Zinc, 3), GT_OreDictUnificator.get(OrePrefixes.dust, Materials.Sulfur, 4), null, null, null, 300, 192);
```
[the remark this answers is not in the document]
Comment (GlodBlock) [reply src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java:537]: the lead zinc solution cost 4x sulfuric acid, 3x galena, 1x sphalerite

File: src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java [hunk 4]
```diff
@@ -2089,88 +2089,127 @@ public void run2() {
 
     private void addKevlarLineRecipes() {
         //Kevlar Line
-        GT_Values.RA.addDistillationTowerRecipe(MaterialsKevlar.DiphenylmethaneDiisocyanateMixture.getFluid(25000L), new FluidStack[]{Materials.HydrochloricAcid.getFluid(4000L)}, GT_OreDictUnificator.get(OrePrefixes.dust, MaterialsKevlar.DiphenylmethaneDiisocyanate, 1L), 2500, 1920);
+        //C15H10N2O2(4HCl) = C15H10N2O2 + 4HCl
```
Comment (miozune) [inline src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java:2092]: C15H10N2O2(5HCl)

File: src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (miozune) [reply src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java:537]: Okay Sulfur is fine. Why 3x Zinc?

File: src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (miozune) [reply src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java:3884]: I mean Stibnite

[the remark this answers is not in the document]
Comment (miozune) [reply src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java:3884]: And some materials don't even need increased output as they have direct smelting recipes (Garnierite, Banded Iron)
Plus 5x Sb2O3 contains 2x Sb, so it should be 2.5x

**Comment by Connor-Colenso** [comment]
Personally I think if this is going to be done we need a reboot way to enforce it in future. Could we write a unit test to ensure that chemical formulas are balanced in future? Or at least those recipes with obvious chemical elements displayed.

File: src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (GlodBlock) [reply src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java]: i mistook lead with zinc...

File: src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java [hunk 5]
```diff
@@ -2089,88 +2089,127 @@ public void run2() {
...
-            GT_Values.RA.addMultiblockChemicalRecipe(new ItemStack[]{GT_Utility.getIntegratedCircuit(18), Materials.SodiumHydroxide.getDust(18), Materials.Manganese.getDust(1)}, new FluidStack[]{new FluidStack(FluidRegistry.getFluid("hydrazine"), 9000)}, new FluidStack[]{Materials.Nitrogen.getGas(9000L), Materials.Water.getFluid(18000L)}, new ItemStack[]{MaterialsKevlar.SodiumHydride.getDust(18)}, 70, 1920);
+            //Too big to fit
+            GT_Values.RA.addMultiblockChemicalRecipe(new ItemStack[]{GT_Utility.getIntegratedCircuit(1), GT_OreDictUnificator.get(OrePrefixes.dust, MaterialsKevlar.RhodiumChloride, 4L), MaterialsKevlar.Triphenylphosphene.getDust(6), MaterialsKevlar.SodiumBorohydride.getDust(2)}, new FluidStack[]{Materials.CarbonMonoxide.getGas(4000L)}, new FluidStack[]{new FluidStack(FluidRegistry.getFluid("phosgene"), 2000), Materials.Water.getFluid(6000L)}, new ItemStack[]{MaterialsKevlar.OrganorhodiumCatalyst.getDust(1)}, 800, 500000);
```
[the remark this answers is not in the document]
Comment (GlodBlock) [reply src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java:2146]: it contains 64+ same items input which gt machine can't process correctly so i leave it unchanged. https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/8166

**Comment by Connor-Colenso** [comment]
Wouldn't it be easier to put a unit test in GT5 to iterate over the Chem reactor hash map and ensure equality unless we have some flag saying not to check it? This way we would make sure future recipes are also balanced.

**Comment by GlodBlock** [comment]
almost impossible to check them automatically. lcr's recipe always using materials from different mods(like bw, gtpp, gt or some random mod), i can get the composition elememt of gt5 material but how to get those of bw and gtpp material? useless all materials are added by gt5 or using a common interface to get their compositions.

File: src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java [hunk 6]
```diff
@@ -2089,20 +2089,20 @@ public void run2() {
...
+            //C13H14N2(HCl) + 2COCl2 = C15H10N2O2(5HCl)
             GT_Values.RA.addMultiblockChemicalRecipe(new ItemStack[]{GT_Utility.getIntegratedCircuit(1)}, new FluidStack[]{MaterialsKevlar.DiaminodiphenylmethanMixture.getFluid(1000L), new FluidStack(FluidRegistry.getFluid("phosgene"), 2000)}, new FluidStack[]{MaterialsKevlar.DiphenylmethaneDiisocyanateMixture.getFluid(1000L)}, null, 600, 1920);
-            //C4H8O + 4H =Pd= C4H10 + H2O
-            GT_Values.RA.addChemicalRecipe(GT_OreDictUnificator.get(OrePrefixes.cell, MaterialsKevlar.Butyraldehyde, 1L), GT_OreDictUnificator.get(OrePrefixes.dustTiny, Materials.Palladium, 1L), Materials.Hydrogen.getGas(4000), new FluidStack(FluidRegistry.getFluid("butanol"), 1000), ItemList.Cell_Empty.get(1L), 200, 30);
-            GT_Values.RA.addMultiblockChemicalRecipe(new ItemStack[]{GT_Utility.getIntegratedCircuit(9), GT_OreDictUnificator.get(OrePrefixes.dust, Materials.Palladium, 1L)}, new FluidStack[]{MaterialsKevlar.Butyraldehyde.getFluid(9000), Materials.Hydrogen.getGas(36000)}, new FluidStack[]{new FluidStack(FluidRegistry.getFluid("butanol"), 9000)}, null, 80, 480);
+            //C4H10O + 2H =Pd= C4H10 + H2O
```
Comment (miozune) [inline src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java:2106]: C4H8O + 2H =Pd= C4H10O

File: src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (miozune) [reply src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java:2149]: I guess it's fixed
https://github.com/GTNewHorizons/GT5-Unofficial/pull/746

File: src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (miozune) [reply src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java:3887]: After second thought, I realized those dusts have standardized electrolysis recipes in the first place, so I believe we should not change these amounts.

File: src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (GlodBlock) [reply src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java:2149]: https://github.com/GTNewHorizons/GT5-Unofficial/blob/master/src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_LargeChemicalReactor.java#L146 lcr is using this code to dismiss the 64+ items. is it ok to remove these code directly if it is fixed?

[the remark this answers is not in the document]
Comment (miozune) [reply src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java:2149]: Ah, didn't know LCR still has that limit. Not 100% sure but I guess it'll work

**Comment by GlodBlock** [comment]
i have to go to bed now, will fix these things tomorrow.

File: src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (GlodBlock) [reply src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java:3887]: chochem suggested to use 'impure ore dust' to distinguish these ore dusts with chemicals so these bounbs recipe can be kept unchanged. i plan to open another pr to deal with it.

[the remark this answers is not in the document]
Comment (miozune) [reply src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java:3886]: @GlodBlock How do you think about this?

[the remark this answers is not in the document]
Comment (GlodBlock) [reply src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java:3886]: wrong ping? it is fine imo

[the remark this answers is not in the document]
Comment (miozune) [reply src/main/java/gregtech/loaders/postload/GT_MachineRecipeLoader.java:3886]: Ah, now I see the intention of these changes, sorry

**Review by miozune — APPROVED** [review]
Let's goooo

**Pull request merged**